tools/read-it-and-keep/trim_reference.py

```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python
+
+from __future__ import print_function
+
+import argparse
+import sys
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('input_file', type=argparse.FileType())
+    parser.add_argument('output_file', type=argparse.FileType('w'), nargs='?', default=sys.stdout)
+    args = parser.parse_args()
+    lines = args.input_file.readlines()
+    i = len(lines) - 1
+    trimmed = False
+    # step backwards through the lines, removing all As until we find a non-A nucleotide
+    while not trimmed:
+        line = lines[i].upper().rstrip()
+        for j in range(len(line) - 1, -1, -1):
+            if line[j] != 'A':
+                lines[i] = line[:j + 1] + '\n'
+                trimmed = True
+                break
+        else:
+            # we processed the whole line - all As - so we don't include this line in the output
+            i -= 1
+    args.output_file.write(''.join(lines[:i + 1]))
```

Comment:
```suggestion
    line_cache = []
    polyA_chars = {'A', 'a', ' '}
    for line in args.input_file:
        line_seq = line.strip()
        if any(c not in polyA_chars for c in line_seq):
            for cached_line in line_cache:
                args.output_file.write(cached_line)
            line_cache = []
            if line_seq[-1] not in polyA_chars:
                args.output_file.write(line)
                continue
        line_cache.append(line)
    # if the cache holds any remaining lines, then the first of these
    # may still contain non-A sequence
    if line_cache:
        rem = line_cache[0].rstrip().rstrip(''.join(polyA_chars))
        if rem:
            rem = rem + '\n'
            args.output_file.write(rem)
```
here's a version that reads input lines one at a time, preserves the non-trimmed input sequence exactly, and does not trip over interior spaces in the sequence.

Reply:
Upps, forgot the header line in my suggestion above :(
Needs a `args.output_file.write(args.input_file.readline())` as the first line, sorry.

Reply:
> As to the script, I think the one I have is good enough, efficient enough for the purpose at hand and easier to understand than the one you wrote, so I'd suggest keeping it as is. I don't think that spaces inside FASTA records are common enough to worry about.

Fair enough :) I guess everyone tends to consider their own code most understandable until 6 months have passed, but you're right that none of my suggested improvements will matter much in the standard case.

Just to note: the minimal change to make your code handle spaces in the sequence, if I'm not mistaken, is:
```
20 -             if line[j] != 'A':
20 +             if line[j] in ['A', ' ']:
```
Though an exotic case, it might not hurt to be prepared for it?

Reply:
Do you mean `if line[j] not in ['A', ' ']:`?

Reply:
Ah, yes of course :)